ddprofiler/app/src/main/java/ddprofiler/analysis/modules/LabelAnalyzer.java

```diff
@@ -0,0 +1,128 @@
+import java.util.List;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import ddprofiler.analysis.TextualDataConsumer;
+import ddprofiler.core.config.ProfilerConfig;
+import xsystem.layers.XStructure;
+import xsystem.learning.LearningModel;
+import xsystem.learning.XStructType;
+
+public class LabelAnalyzer implements TextualDataConsumer {
+
+    final private Logger LOG = LoggerFactory.getLogger(LabelAnalyzer.class.getName());
+
+    private ProfilerConfig pc;
+    private double scoreThreshold;
+    private static ArrayList<XStructType> xStructureReference = null;
+    private String label;
+
+    public LabelAnalyzer(ProfilerConfig pc) {
+        this.pc = pc;
+        this.scoreThreshold = pc.getDouble("xsystem.threshold");
+
+        if (LabelAnalyzer.xStructureReference == null && pc.getBoolean("xsystem")) {
+            String referenceFilePath = pc.getString("xsystem.reference");
+            LabelAnalyzer.xStructureReference = (new LearningModel()).readXStructsfromJSON(referenceFilePath);
+        } else {
+            LOG.warn("Reference file already initialized or XSystem is not enabled");
+        }
+    }
+
+    /**
+     * For testing
+     * @param string
+     */
+    private LabelAnalyzer(String string) {
+        String referenceFilePath = string;
+        LabelAnalyzer.xStructureReference = (new LearningModel()).readXStructsfromJSON(referenceFilePath);
+        scoreThreshold = 0.5;
+    }
+
+    @Override
+    public boolean feedTextData(List<String> records) {
+        if (records == null || records.isEmpty()) {
+            LOG.warn("No records to analyze");
+        }
+        if (LabelAnalyzer.xStructureReference == null) {
+            LOG.warn("Reference file not initialized");
+        }
+        if (pc.getBoolean("xsystem")) {
+            this.label = labelListOfStrings((ArrayList<String>) records);
+        } else {
+            LOG.info("XSystem is not enabled");
+        }
+        return true;
+    }
+
+    private String labelListOfStrings(ArrayList<String> strings) {
+        String label = null;
+        ArrayList<Double> scoreList = new ArrayList<>();
+        XStructure toBeLabeled = (new XStructure()).addNewLines(strings);
+
+        for (XStructType struct : xStructureReference) {
+            double score = struct.xStructure.compareTwo(toBeLabeled, struct.xStructure);
+            scoreList.add(score);
+        }
+        double maxScore = Collections.max(scoreList);
+        int maxIndex = scoreList.indexOf(maxScore);
+
+        if (maxScore >= scoreThreshold) {
+            label = xStructureReference.get(maxIndex).type;
+            return label;
+        }
+
+        return label;
+    }
+
+    public static void nullifyXStructureReference() {
+        LabelAnalyzer.xStructureReference = null;
+    }
+
+    public String getLabel() {
+        return label;
+    }
+
+    public static void main(String[] args) {
+
+        LabelAnalyzer autolabel = new LabelAnalyzer(
```

Comment:
Could Xsystem deal with more complex/dirty spatio-temporal data types(formats)? For example, when the date uses a different format(e.g. 2023-05-30 00:00:00) or coordinates have the format like "Point(12.313414414, -31.314141414)" or spatial attributes are not of coordinate format (e.g. zipcode, street address, state/city name, etc..). I suggest that you can write a short document about what data types xsystem can handle right now and add more tests.

Reply:
I just moved the test to a new file dedicated to unit tests and added a few more. I just recently added support for zip code. Therefore, date & time, location (coordinates), and zip code are currently supported.

The automated labeling module depends on two main things, the reference file and a similarity threshold. The reference file is a collection of labels and their known pattern (XStructure). The threshold decides whether an attribute fits the label when compared with said pattern.

Currently, there are three main problems with the current implementation (assuming dirty inputs):
- The desired label of an attribute doesn't exist in the reference file. This problem can be fixed by adding a reference for the related label.
- The attribute has a lower similarity score with the desired label than the defined similarity threshold. Lowering the threshold can fix this problem.
- An attribute has a higher similarity score with an undesired label than the desired label. This one is tricky, and I don't have a robust way to fix it yet.

Finding a suitable threshold is difficult, and different reference file may be needed for different user requirements, so I also allow the user to set their similarity threshold and use their reference file from the profiler config.